yagb-core/src/emulator/ppu/ppu-cgb.ts

```diff
@@ -389,6 +405,101 @@ export class PpuCgb extends PpuBase {
         }
     };
 
+    private invalidRead: ReadHandler = (address) => {
+        this.system.log(`Invalid read from HDMA register ${hex16(address)}`);
+
+        return 0x00;
+    };
+
+    private hdma1Write: WriteHandler = (_, value) => {
+        this.hdmaSource = (this.hdmaSource & 0xff) | (value << 8);
+        this.hdmaSource &= 0xfff0;
+    };
+
+    private hdma2Write: WriteHandler = (_, value) => {
+        this.hdmaSource = (this.hdmaSource & 0xff00) | (value & 0xf0);
+        this.hdmaSource &= 0xfff0;
+    };
+
+    private hdma3Write: WriteHandler = (_, value) => {
+        this.hdmaDestination = (this.hdmaDestination & 0xff) | (value << 8);
+        this.hdmaDestination &= 0x1ff0;
+    };
+
+    private hdma4Write: WriteHandler = (_, value) => {
+        this.hdmaDestination = (this.hdmaDestination & 0xff00) | (value & 0xf0);
+        this.hdmaDestination &= 0x1ff0;
+    };
+
+    private hdma5Read: ReadHandler = () => {
+        if (this.hdmaLength === 0) return 0xff;
+
+        return (this.hdmaOn ? 0x00 : 0x80) | ((this.hdmaLength >> 4) - 1);
+    };
+
+    private hdma5Write: WriteHandler = (_, value) => {
+        if (this.hdmaOn && TransferMode.hblank === this.hdmaMode) {
+            if ((value & 0x80) === 0) {
+                this.hdmaOn = false;
+                return;
+            }
+        }
+
+        if (this.hdmaOn) {
+            this.system.log('HDMA transfer already in progress!');
+
+            return;
+        }
+
+        this.hdmaOn = true;
+        this.hdmaMode = value & 0x80;
+        this.hdmaLength = ((value & 0x7f) + 1) << 4;
+
+        if (this.hdmaMode !== TransferMode.general) return;
+
+        this.hdmaCopyMemory(this.hdmaLength);
+
+        this.resetHDMA();
+    };
+
+    private hdmaCopyMemory(length: number) {
+        const destination = 0x8000 + this.hdmaDestination;
+
+        let value;
+        for (let address = 0; address < length; address++) {
+            if (0x9ff0 < destination + address) {
+                this.system.log('Address overflow during HDMA transfer');
```

Comment:
I am pretty sure that what actually happens is that the address wraps back to the beginning of VRAM. Maybe do that instead, and keep the log while developmet

Reply:
From Pandoc:
If the transfer’s destination address overflows, the transfer stops prematurely. The status of the registers if this happens still needs to be investigated.
https://gbdev.io/pandocs/CGB_Registers.html#bit-7--1--hblank-dma